packages/contracts-bedrock/src/safe/TimelockGuard.sol

```diff
@@ -0,0 +1,540 @@
+
+    /// @notice Called by the Safe before executing a transaction
+    /// @dev Implementation of IGuard interface
+    function checkTransaction(
+        address _to,
+        uint256 _value,
+        bytes memory _data,
+        Enum.Operation _operation,
+        uint256 _safeTxGas,
+        uint256 _baseGas,
+        uint256 _gasPrice,
+        address _gasToken,
+        address payable _refundReceiver,
+        bytes memory,
+        address
+    )
+        external
+        override
+    {
+        Safe callingSafe = Safe(payable(msg.sender));
+
+        if (_safeState[callingSafe].timelockDelay == 0) {
+            // We return immediately. This is important in order to allow a Safe which has the
+            // guard set, but not configured to complete the setup process.
+            // It is also just a reasonable thing to do, since an unconfigured Safe must have a
+            // delay of zero.
+            return;
+        }
+
+        // Get the nonce of the Safe for the transaction being executed,
+        // since the Safe's nonce is incremented before the transaction is executed,
+        // we must subtract 1.
+        uint256 nonce = callingSafe.nonce() - 1;
+
+        // Get the transaction hash from the Safe's getTransactionHash function
+        bytes32 txHash = callingSafe.getTransactionHash(
+            _to, _value, _data, _operation, _safeTxGas, _baseGas, _gasPrice, _gasToken, _refundReceiver, nonce
+        );
+
+        // Get the scheduled transaction
+        ScheduledTransaction storage scheduledTx = _safeState[callingSafe].scheduledTransactions[txHash];
+
+        // Check if the transaction was cancelled
+        if (scheduledTx.cancelled) {
+            revert TimelockGuard_TransactionAlreadyCancelled();
+        }
+
+        // Check if the transaction has been scheduled
+        if (scheduledTx.executionTime == 0) {
+            revert TimelockGuard_TransactionNotScheduled();
+        }
+
+        // Check if the timelock delay has passed
+        if (scheduledTx.executionTime > block.timestamp) {
+            revert TimelockGuard_TransactionNotReady();
+        }
+
+        // Check if the transaction has already been executed
+        // Note: this is of course enforced by the Safe itself, but we check it here for
+        // completeness
+        if (scheduledTx.executed) {
+            revert TimelockGuard_TransactionAlreadyExecuted();
+        }
+
+        // Set the transaction as executed
+        scheduledTx.executed = true;
+        _safeState[callingSafe].pendingTxHashes.remove(txHash);
+
+        // Reset the cancellation threshold
+        _resetCancellationThreshold(callingSafe);
+
+        emit TransactionExecuted(callingSafe, nonce, txHash);
+    }
+
+    /// @notice Called by the Safe after executing a transaction
+    /// @dev Implementation of IGuard interface
+    function checkAfterExecution(bytes32, bool) external override {
+        // Do nothing
```

Comment:
sanity check: there is no case where the Safe execution might run out of gas and the transaction sigs can be replayed? if there is, we'd want to move the state changing methods/event emission into this method. I don't think there is, but just confirming

Reply:
I'd thought a bit about this, and is generally a trade off with this question of letting something remain executable after an OOG revert.

But now that you've mentioned it, it's pretty clear that we should follow the same behaviour of the Safe, which is that the nonce is not incremented if the call is not successfully (at least in the [default case](https://github.com/safe-global/safe-smart-account/blob/f3af2a9a24aa47faea3dbd8ebcdf665694975a0f/contracts/GnosisSafe.sol#L180) where `gasPrice == safeTxGas == 0`). So we should follow the safe and only call it executed if `success == true`.

I'll do that.